More robust setting of kTopC

jm-c opened the pull request:
## What changes does this PR introduce?
1. Adjust changes from PR #465 by placing the code that deals with the update of `kTopC` within "#ifdef ALLOW_SHELFICE_REMESHING" and enable to turn this off with new run-time parameter `SHI_update_kTopC`.
2. Change the initial setting of `kTopC` to be function of `shelficeMassInit` when using `SHELFICEMassStepping` and make it more robust (fix issue #798) in the other case.

## What is the current behaviour? 
- initial `kTopC` setting issue #798
- update of kTopC code requires special care for AD processing

## What is the new behaviour 
- issue is fixed ;
- also more consistent initialisation of kTopC when ice-shelf mass is evolving.
- kTopC update code is protected with CCP option (simpler for AD if not used), and a run-time switch allows to turn it off.
- this new run-time switch  `SHI_update_kTopC` allows to refine/improve the parameter consistency check

## Does this PR introduce a breaking change? 
No

## Other information:
Also:
- reset R_shelfIce and shelficeMassInit to zero where kTopC=0
- allow to print (currently with debugLevel >= 3) list of location with potential mis-match between kTopC and shelficeMassInit or R_shelfIce.

## Suggested addition to `tag-index`
o pkg/shelfice:
  - put the update of kTopC code within "#ifdef ALLOW_SHELFICE_REMESHING/#endif"
    and add run-time switch "SHI_update_kTopC" to allow to turn it off.
  - Change the initial setting of "kTopC" to remain unaffected by machine
    truncation error and if using SHELFICEMassStepping, set it according to
    shelficeMassInit.

jm-c commented:
Note: 
1. this is not fully consistent with changes from PR #465 and will need to be sorted out:
In short, PR #465 was intended to fix issue #379 which has been raised for shelfice-remeshingi case, but was made more general by applying the "kTopC" update anytime `SHELFICEMassStepping` is used.
So the case that is less clear is when `SHELFICEMassStepping`  is used without shelfice-remeshing  and "kTopC" is updated during the simulation.

2. the changes here will address @dngoldberg comment (in issue #379, on Oct 08, 2020):
> In the second, we will need care not to initialise Ro_surf to rF(k) -- even though after remeshing happens this will be the case. But this is not quite so robust.

so that, when using shelfice-remeshing, `R_shelfIce` could now be initilized to zero, kTopC will be set > 0 if ` shelficeMass` > 0.

mjlosch reviewed (approved):
This PR solves both issues raised in #798, also for the complicated configuration where @nicolasdettling found it initially.

jm-c commented:
I went a little bit further in my thinking about the update of kTopC  and would like to revisit/adjust the changes made in PR #465 for the case where `SHELFICEMassStepping`=T but without remeshing (i..e., `SHELFICEremeshFrequency` = 0):

1. with linear Free-Surface, since the ocean `hFac` are fixed in time, updating `kTopC` will result in inconsistent domain (with a "hole" after iceshelf vanishes or overlapping of iceshelf over ocean after iceshelf spreads). So I would prefer to disable it.
2. I would like to braket the update of `kTopC` code within "#ifdef ALLOW_SHELFICE_REMESHING / #endif" since it also affects the ice-shelf extension (even through there is no true remeshing involded when updating kTopC). This would make the code simpler for the adjoint. Also  the current `pkg/seaice` would work with shelfice default setting  (i.e,, without kTopC update).
3. and adding a run-time switch would make code more clear, easier to have a check for pkg/seaice, and easier to adjust the logic (in case I changed my mind again).

And regarding the changes in this branch (my review):

1. the changes here in `shelfice_read_pickup.F` + the "remeshing restart case" in `shelfice_init_fixed.F` are not necessary since, for this case, `kTopC` will be reset in `shelfice_init_varia.F`.
2. and the "remeshing starting case" can be used any time `SHELFICEMassStepping` is used (I think in this case it's preferable to set an intial ice-shelf mass field rather than using the `SHELFICEloadAnomalyFile` setting).

jm-c commented:
The last commit (to avoid un-used variables in `pkg/obcs` ) should have been part of PR #579 (but forgot to add these files) and is unrelated to "kTopC" apart from getting all these "Unused variable" warnings (37 in total)  in experiment `shelfice_2d_remesh` which I use to check these "tTopC" changes.

jm-c commented:
I removed the "work-in-progress" label and I've just updated this PR description. 
But since I made significant changes from the first reviewed version, it would be useful if  @mjlosch could check the latest version (I am not planning to make other changes now). Thanks.

mjlosch commented on pkg/shelfice/shelfice_check.F:
```diff
@@ -268,15 +293,14 @@ SUBROUTINE SHELFICE_CHECK( myThid )
-     &                      SQUEEZE_RIGHT, myThid )
+        CALL PRINT_ERROR( msgBuf, myThid )
+        WRITE(msgBuf,'(2A)') 'SHELFICE_CHECK: ',
+     &   ' does not work with SHI_update_kTopC=T'
```
```suggestion
     &   'and does not work with SHI_update_kTopC=T'
```

jm-c replied:
I've just change it as you suggested. Thanks.

mjlosch reviewed (approved):
changes look good to me. One minor comment.

jm-c commented:
I think this PR is ready to be merged. Unless someone wants to add something, will merge it soon, like later today.
@mjlosch do you agree ?
